src/cargo/core/resolver/dep_cache.rs

```diff
@@ -267,15 +282,16 @@ impl<'a> RegistryQueryer<'a> {
         // which can satisfy that dependency.
         let mut deps = deps
             .into_iter()
-            .map(|(dep, features)| {
-                let candidates = self.query(&dep).chain_err(|| {
+            .filter_map(|(dep, features)| match self.query(&dep) {
+                Ok(Poll::Ready(candidates)) => Some(Ok((dep, candidates, features))),
+                Ok(Poll::Pending) => None, // we can ignore Pending deps, resolved will be repeatedly called until there are none to ignore
```

Comment:
This actually feels like it's somewhat dangerous to ignore because we're populating a `summary_cache` with results that will get invalidated later. Although if the entire resolver is thrown away that's less of an issue.

Reply:
So the nice thing about skipping the `dependency` (by doing the skipping here) is that if we get a contradiction (proof that there is nothing that solves all the requirements) then we know that that proof will still be valid when more `dependency` come in. So the outer loop around `activate_deps_loop` can use `?` to fail fast when we find a problem.

At the moment we are throwing away the entire resolver. (Make it right before we make it fast.) If we want to be less wasteful, we can keep the parts of `registry_cache` that are not pending. We can even make some side table to record which parts of the `summary_cache` need to be removed, so we can keep the rest.

Reply:
That's a good point yeah. I think we should stick to throwing away the entire resolver for now since it's the most conservative thing to do. If we hit performance issues we could consider preserving some of it, but we should probably investigate which are the most important parts to preserve for that.

Reply:
Work had a zoom holiday party, and so you get a commit that invalidates these caches more precisely. `registry_cache` and `summary_cache` keep track of whether there from a `Pending` response so that `reset_pending` can remove only the needed ones.